[eclipse/xtext#1425] show number of matches in ReferenceSearchViewPage

riederm opened the pull request:
The Find-References page now shows number of matches as a suffix in the
label that describes the search result.
The label is refreshed when the search finished - not after every single
hit was added to search result to avoid extensive redraws.

Signed-off-by: Mathias Rieder <mathias.rieder@gmail.com>

cdietrich commented:
Nice. I wonder if we should update the match count when matches are removed from the result as it is done for Java.

riederm commented:
Hey Christian,
of course you're right. I took a closer look at the whole remove-semantics and I was playing around with the code now for several evenings and I'm having a hard time to get this working without breaking a bunch of compatibilities. This has mainly two reasons:

1) the original code uses a ReferenceSearchResult that holds a list of IReferenceDescriptions - thats fine. IReferenceDescriptions are added to the ReferenceSearchResult and the viewer adds corresponding objects to the tree via the ReferenceSearchResultContentProvider receiving events from the search-result. So far so good - removing from the viewer is accomplished via the content-provider's internal model (ReferenceSearchViewTreeNode) - the origina SearchResult model is not touched. This has two implications: 

a) the label that shows the number of matches does not know anything about the internal model managed by the contentprovider, so from the outside the ReferenceSearchResult still contains the 
same number of search result.

b) the SearchResult stored in the history (Show previous Searches) still holds the removed matches.

so I was going to do the removes on the ReferenceSearchResult and propagate the removes again via event into the contentprovider. I also feel that the responsibilities are cleaner that way - it just means some serious changes to the contentprovider - I'm not sure if you want such API breaking changes in this ContentProvider. (e.g. some protected methods make no sense anymore, it would no longer implement the IResourceDescription.Event.Listener since the removes are propagated via the ReferenceSearchResult-

2) The second problem is the more serious problem. There are two ways hot elements are removed from the table - either the user deletes a selected item, or an item is removed because of an index-change-event and we can no longer find that particular reference. The thing is, I cannot make this index-based feature work in the original code. 

The thing is that the ReferenceFinder gives me quite different source-URIs than what the IndexChangeEvent would provide me:

- platform:/resource/org.xtext.example.mydsl/src/org/xtext/example/mydsl/MyDsl.xtext#/0/@rules.0/@type
- platform:/resource/org.xtext.example.mydsl/src/org/xtext/example/mydsl/MyDsl.xtext#/0/@rules.1/@type

 The ReferenceFinder gave me these sourceURIs:
- platform:/resource/org.xtext.example.mydsl/src/org/xtext/example/mydsl/MyDsl.xtext#//@rules.0/@alternatives/@terminal
- platform:/resource/org.xtext.example.mydsl/src/org/xtext/example/mydsl/MyDsl.xtext#//@rules.1/@alternatives/@terminal

so these URIs are quite different. The content-index in the beginning may differ, and I dont really see why the fragment would be so different? Do you have any clue for me? It seams that it has something to do with local references?

cdietrich commented:
I will see when I find the time to look at this.
Maybe @kthoms or @szarnekow  have an idea too

cdietrich commented:
hmmm i assume for the change events we need to start a new query.
they just give you which named objects changed but not which reference holder changed.

riederm commented:
I agree, it looks like this is necessary. This changes the feature quite a bit in a way that it not only removes invalid hits, but it would also add new hits as soon as they're built - so quite some activity in the search-view. 
do we want this?

cdietrich commented:
hmmm. am not sure how other searches e.g file and java behave.
if i remember correct they dont live update. you have to rerun.
but you can delete find results and the number updates.

@kthoms @szarnekow what do you two think?

szarnekow commented:
I don't think that we can easily make this a live view of the search results. Having this as a snapshot that updates the counter when you remove entries from the view itself (maybe adding a remark that n items have been removed from the view) sounds reasonable.

riederm commented:
> I don't think that we can easily make this a live view of the search results. Having this as a snapshot that updates the counter when you remove entries from the view itself (maybe adding a remark that n items have been removed from the view) sounds reasonable.

So do i get this right, that I only handle removes by the user and ignore index-changes at all? There is existing code that listens for index-changes in order to remove invalidated entries, but it does not work due to the URI-differences I explained above. Maybe it worked in earlier versions? i don't know.

cdietrich commented:
yes. hust handle the deletes by the user would be perfectly fine for me. if a use wants to update the search to model changes, then the search can be rexecuted. 

java search behaves the same. if you want updated results you have to reexecute the search

cdietrich commented:
i wonder how to continue with this pr

riederm commented:
my apologies for beeing so silent, i was pulled into work the last couple of weeks. I'll give this commit another try next week.

riederm commented:
i gave this another visit. the state of the current version is:

- number of matches are shown as requested
- removing matches updates the number of matches
- removed matches are remembered by the SearchResult so they are not added again when the query is re-run form the view's history (I find it a weird behavior, but Java acts the same, so I implemented it accordingly)

- live changes from the index are still not reflected by the search result. I cannot make the existing code work due to the different URIs as explained earlier. (do you want me to remove the old code that does not work?)

cdietrich commented:
Yes we should keep it minimal and working (as in java)

cdietrich commented:
@riederm loooks like there are compile errors now against the Oxygen target platform

riederm commented:
> @riederm loooks like there are compile errors now against the Oxygen target platform

whoops, some java11 methods slipped in

cdietrich commented:
@kthoms @ArneDeutsch what do you think?

cdietrich commented:
ping @kthoms @ArneDeutsch

cdietrich commented:
@kthoms what do you think?

kthoms commented on org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/findrefs/ReferenceSearchResultContentProvider.java:
```diff
@@ -197,10 +199,28 @@ public IStatus runInUIThread(IProgressMonitor monitor) {
+	
+
+	private void expandToFirstChild() {
+		if (rootNodes != null && !rootNodes.isEmpty()){
```
personally I prefer fast returns instead of an if-block that spans the whole method body

cdietrich commented:
thx for the contribution @riederm
